src/util/indexrange.h

```diff
@@ -0,0 +1,175 @@
+#ifndef MIXXX_INDEXRANGE_H
+#define MIXXX_INDEXRANGE_H
+
+#include <QtDebug>
+
+#include <iosfwd>
+#include <utility>
+
+#include "util/assert.h"
+#include "util/types.h"
+
+
+namespace mixxx {
+
+// A half-open, directed range of indices with limited mutability.
+class IndexRange final: private std::pair<SINT, SINT> {
+    typedef std::pair<SINT, SINT> Super;
+
+  public:
+    using Super::swap;
+
+    IndexRange()
+        : Super(0, 0) {
+        DEBUG_ASSERT(empty());
+    }
+
+    static IndexRange between(SINT head, SINT tail) {
+        return IndexRange(head, tail);
+    }
+    static IndexRange forward(SINT head, SINT length) {
+        DEBUG_ASSERT(length >= 0);
+        return IndexRange(head, head + length);
+    }
+    static IndexRange backward(SINT head, SINT length) {
+        DEBUG_ASSERT(length >= 0);
+        return IndexRange(head, head - length);
+    }
+
+    // The near boundary (inclusive)
+    SINT head() const {
+        return first;
+    }
+    // The opposite boundary (exclusive)
+    SINT tail() const {
+        return second;
+    }
+
+    bool empty() const {
+        return head() == tail();
+    }
+
+    enum class Orientation {
+        Empty,
+        Forward,
+        Backward,
+    };
```

Comment:
I am not entirely through with my reading...
I just wonder if we need a range class that supports forward and backward in the same time. Will the using code simpler or not save if we have dedicated classes or sub classes only supporting one direction?